src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.java

```
@@ -196,27 +198,32 @@ public ParcelFileDescriptor openDocument(String documentId, String mode, Cancell
 
-        if (ocFile.isDown()) {
-            RemoteOperationResult result;
-            try {
-                result = new SynchronizeFileOperation(ocFile, null, user, true, context)
-                    .execute(document.getClient(), document.getStorageManager());
-            } catch (Exception e) {
-                throw getFileNotFoundExceptionWithCause("Error synchronizing file: " + ocFile.getFileName(), e);
-            }
-            if (result.getCode() == RemoteOperationResult.ResultCode.SYNC_CONFLICT) {
-                // TODO show a conflict notification with a pending intent that shows a ConflictResolveDialog
-                Log_OC.w(TAG, "Conflict found: " + result);
-            } else if (!result.isSuccess()) {
+        int accessMode = ParcelFileDescriptor.parseMode(mode);
+
+        boolean needsDownload = accessMode != ParcelFileDescriptor.MODE_WRITE_ONLY;
+        if (needsDownload && ocFile.isDown()) {
+            RemoteOperationResult result = new ReadFileRemoteOperation(ocFile.getRemotePath())
+                .execute(document.getClient());
+            if (result.isSuccess()) {
+                OCFile serverFile = FileStorageUtils.fillOCFile((RemoteFile) result.getData().get(0));
+                boolean serverChanged = !serverFile.getEtag().equals(ocFile.getEtag());
+                boolean localChanged = ocFile.getLocalModificationTimestamp() > ocFile.getLastSyncDateForData();
+                if (localChanged && serverChanged) {
+                    // TODO show a conflict notification with a pending intent that shows a ConflictResolveDialog
+                    Log_OC.w(TAG, "Conflict found: " + result);
+                    // open local version for now
+                    needsDownload = false;
+                } else {
+                    needsDownload = serverChanged;
+                }
+            } else if (result.getCode() == RemoteOperationResult.ResultCode.FILE_NOT_FOUND) {
```

What if the result was not a success, but had a different error than this one? Are we just silently ignoring it?

---

We then should exit with any valid return code, 404?

---

> Did you run the integration tests I had added? (The CI is flaky unfortunately)
Yes

> Would maybe be nice to add another test to them that shows the improved behavior, aka not serving stale files anymore?
Done